File: backend/algorithms.py
import heapq
from datetime import datetime
import time
from typing import List, Dict, Any, Union

def parse_iso(ts: Union[str, datetime]) -> Union[datetime, str]:
    try:
        if ts.endswith('Z'):
            ts = ts[:-1] + '+00:00'
        return datetime.fromisoformat(ts)
    except Exception:
        return ts

def k_way_merge(arrays: List[List[Dict[str, Any]]]) -> Dict[str, Any]:
    result = []
    steps = []
    heap = []
    
    for idx, arr in enumerate(arrays):
        if len(arr) > 0:
            item = arr[0]
            ts = parse_iso(item.get('timestamp', ''))
            heapq.heappush(heap, (ts, idx, 0, item))
            
    steps.append({
        'type': 'init',
        'desc': f'Initialized min-heap with {len(heap)} elements (one from each active satellite)',
        'heapState': [f"[{h[3].get('satellite')}] {h[3].get('timestamp')}" for h in heap[:8]],
        'extracted': None
    })
    
    iteration = 0
    while heap:
        ts, srcIdx, itemIdx, min_item = heapq.heappop(heap)
        clean = min_item.copy()
        result.append(clean)
        
        nextIdx = itemIdx + 1
        if nextIdx < len(arrays[srcIdx]):
            next_item = arrays[srcIdx][nextIdx]
            next_ts = parse_iso(next_item.get('timestamp', ''))
            heapq.heappush(heap, (next_ts, srcIdx, nextIdx, next_item))
            
        if len(steps) < 30:
            steps.append({
                'type': 'extract',
                'desc': f"Extracted [{min_item.get('satellite')}] @ {min_item.get('timestamp')} -> result[{len(result)-1}]",
                'heapState': [f"[{h[3].get('satellite')}] {h[3].get('timestamp')}" for h in heap[:8]],
                'extracted': min_item.get('satellite')
            })
        iteration += 1
        
    steps.append({
        'type': 'done',
        'desc': f'✅ Merge complete! {len(result)} records unified in chronological order',
        'heapState': [],
        'extracted': None
    })
    return {'result': result, 'steps': steps}

def heap_merge(arrays: List[List[Dict[str, Any]]]) -> Dict[str, Any]:
    heap = []
    steps = []
    total_inserted = 0
    for idx, arr in enumerate(arrays):
        for item in arr:
            ts = parse_iso(item.get('timestamp', ''))
            heapq.heappush(heap, (ts, idx, item))
            total_inserted += 1
            
        sat_name = arr[0].get('satellite') if arr else 'Unknown'
        steps.append({
            'type': 'insert',
            'desc': f'Inserted all {len(arr)} records from {sat_name} into priority queue (total: {total_inserted})',
            'heapState': [f"[{h[2].get('satellite')}] {h[2].get('timestamp')}" for h in heap[:8]],
            'extracted': sat_name
        })
        
    result = []
    while heap:
        ts, idx, min_item = heapq.heappop(heap)
        result.append(min_item)
        if len(steps) < 30:
            steps.append({
                'type': 'extract',
                'desc': f"Extracted min: [{min_item.get('satellite')}] @ {min_item.get('timestamp')}",
                'heapState': [f"[{h[2].get('satellite')}] {h[2].get('timestamp')}" for h in heap[:8]],
                'extracted': min_item.get('satellite')
            })
            
    steps.append({
        'type': 'done',
        'desc': f'✅ Heap merge complete! {len(result)} records processed',
        'heapState': [],
        'extracted': None
    })
    return {'result': result, 'steps': steps}

def naive_merge(arrays: List[List[Dict[str, Any]]]) -> Dict[str, Any]:
    steps = []
    combined = []
    for arr in arrays:
        combined.extend(arr)
        sat_name = arr[0].get('satellite') if arr else 'Unknown'
        steps.append({
            'type': 'concat',
            'desc': f'Concatenated {len(arr)} records from {sat_name} -- running total: {len(combined)}',
            'heapState': [],
            'extracted': sat_name
        })
        
    steps.append({
        'type': 'sort',
        'desc': f'Sorting {len(combined)} records using Python sort (Timsort) -- O(N log N)',
        'heapState': [],
        'extracted': None
    })
    result = sorted(combined, key=lambda x: parse_iso(x.get('timestamp', '')))
    steps.append({
        'type': 'done',
        'desc': f'✅ Naive merge complete! {len(result)} records sorted chronologically',
        'heapState': [],
        'extracted': None
    })
    return {'result': result, 'steps': steps}

def run_merge(arrays: List[List[Dict[str, Any]]], algorithm: str) -> Dict[str, Any]:
    start = time.perf_counter()
    if algorithm == 'heap':
        output = heap_merge(arrays)
    elif algorithm == 'naive':
        output = naive_merge(arrays)
    else:
        output = k_way_merge(arrays)
    elapsed = round((time.perf_counter() - start) * 1000, 3)
    output['elapsed'] = elapsed
    output['algorithm'] = algorithm
    return output

File: algorithms.js
/**
 * algorithms.js — Merge Algorithm Implementations
 * K-Way Merge Sort, Heap Merge, Naive Merge
 * All with step-by-step visualization tracking
 */

// ──────────────────────────────────────────────
// Min-Heap Implementation
// ──────────────────────────────────────────────
class MinHeap {
  constructor() { this.heap = []; }

  insert(item) {
    this.heap.push(item);
    this._bubbleUp(this.heap.length - 1);
  }

  extractMin() {
    if (this.heap.length === 0) return null;
    const min = this.heap[0];
    const last = this.heap.pop();
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this._sinkDown(0);
    }
    return min;
  }

  _bubbleUp(i) {
    while (i > 0) {
      const parent = Math.floor((i - 1) / 2);
      const parentTime = this.heap[parent]._time || new Date(this.heap[parent].timestamp).getTime();
      const iTime = this.heap[i]._time || new Date(this.heap[i].timestamp).getTime();
      if (parentTime <= iTime) break;
      [this.heap[parent], this.heap[i]] = [this.heap[i], this.heap[parent]];
      i = parent;
    }
  }

  _sinkDown(i) {
    const n = this.heap.length;
    while (true) {
      let smallest = i;
      const left = 2 * i + 1, right = 2 * i + 2;
      
      if (left < n) {
        const leftTime = this.heap[left]._time || new Date(this.heap[left].timestamp).getTime();
        const smallestTime = this.heap[smallest]._time || new Date(this.heap[smallest].timestamp).getTime();
        if (leftTime < smallestTime) smallest = left;
      }
      
      if (right < n) {
        const rightTime = this.heap[right]._time || new Date(this.heap[right].timestamp).getTime();
        const smallestTime = this.heap[smallest]._time || new Date(this.heap[smallest].timestamp).getTime();
        if (rightTime < smallestTime) smallest = right;
      }

      if (smallest === i) break;
      [this.heap[i], this.heap[smallest]] = [this.heap[smallest], this.heap[i]];
      i = smallest;
    }
  }

  size() { return this.heap.length; }
  peek() { return this.heap[0] || null; }
  toArray() { return [...this.heap]; }
}

// ──────────────────────────────────────────────
// K-Way Merge Sort (Min-Heap Based)
// ──────────────────────────────────────────────
function kWayMerge(arrays) {
  const heap = new MinHeap();
  const pointers = new Array(arrays.length).fill(0);
  const result = [];
  const steps = [];

  // Pre-parse timestamps for performance and reliability
  const preParsedArrays = arrays.map(arr => 
    arr.map(item => ({ ...item, _time: new Date(item.timestamp).getTime() }))
       .sort((a, b) => a._time - b._time)
  );

  // Initialize heap with first element of each array
  preParsedArrays.forEach((arr, idx) => {
    if (arr.length > 0) {
      heap.insert({ ...arr[0], _srcIdx: idx });
    }
  });

  steps.push({
    type: 'init',
    desc: `Initialized min-heap with ${heap.size()} elements (one from each satellite)`,
    heapState: heap.toArray().map(x => `[${x.satellite}] ${new Date(x.timestamp).toLocaleTimeString()}`),
    extracted: null,
    columns: preParsedArrays.map((arr, i) => ({ sat: arr[0]?.satellite || `Array${i}`, ptr: 0, len: arr.length })),
  });

  while (heap.size() > 0) {
    const min = heap.extractMin();
    const srcIdx = min._srcIdx;
    pointers[srcIdx]++;

    const clean = { ...min };
    delete clean._srcIdx;
    delete clean._time; // Remove internal tracking property
    result.push(clean);

    // Always advance the heap with next element from the same source
    if (preParsedArrays[srcIdx][pointers[srcIdx]]) {
      heap.insert({ ...preParsedArrays[srcIdx][pointers[srcIdx]], _srcIdx: srcIdx });
    }

    // Track up to 30 steps for visualization
    if (steps.length < 30) {
      steps.push({
        type: 'extract',
        desc: `Extracted [${min.satellite}] @ ${new Date(min.timestamp).toLocaleTimeString()} → result[${result.length - 1}]`,
        heapState: heap.toArray().slice(0, 8).map(x => `[${x.satellite}] ${new Date(x.timestamp).toLocaleTimeString()}`),
        extracted: min.satellite,
        columns: preParsedArrays.map((arr, i) => ({ sat: arr[0]?.satellite || `Array${i}`, ptr: pointers[i], len: arr.length })),
      });
    }
  }

  steps.push({
    type: 'done',
    desc: `✅ Merge complete! ${result.length} records unified in chronological order`,
    heapState: [],
    extracted: null,
    columns: [],
  });

  return { result, steps };
}

// ──────────────────────────────────────────────
// Heap Merge (Similar but tracks differently)
// ──────────────────────────────────────────────
function heapMerge(arrays) {
  const heap = new MinHeap();
  const steps = [];

  // Pre-parse and sort sources
  const preParsedArrays = arrays.map(arr => 
    arr.map(item => ({ ...item, _time: new Date(item.timestamp).getTime() }))
       .sort((a, b) => a._time - b._time)
  );

  let totalInserted = 0;
  preParsedArrays.forEach((arr, idx) => {
    arr.forEach(item => {
      heap.insert({ ...item, _srcIdx: idx });
      totalInserted++;
    });
    steps.push({
      type: 'insert',
      desc: `Inserted all ${arr.length} records from ${arr[0]?.satellite} into priority queue (total: ${totalInserted})`,
      heapState: heap.toArray().slice(0, 8).map(x => `[${x.satellite}] ${new Date(x.timestamp).toLocaleTimeString()}`),
      extracted: arr[0]?.satellite,
      columns: [],
    });
  });

  const result = [];
  while (heap.size() > 0) {
    const min = heap.extractMin();
    const clean = { ...min };
    delete clean._srcIdx;
    delete clean._time;
    result.push(clean);

    if (steps.length < 30) {
      steps.push({
        type: 'extract',
        desc: `Extracted min: [${min.satellite}] @ ${new Date(min.timestamp).toLocaleTimeString()}`,
        heapState: heap.toArray().slice(0, 8).map(x => `[${x.satellite}] ${new Date(x.timestamp).toLocaleTimeString()}`),
        extracted: min.satellite,
        columns: [],
      });
    }
  }

  steps.push({
    type: 'done',
    desc: `✅ Heap merge complete! ${result.length} records processed`,
    heapState: [],
    extracted: null,
    columns: [],
  });

  return { result, steps };
}

// ──────────────────────────────────────────────
// Naive Merge (Concatenate + Sort)
// ──────────────────────────────────────────────
function naiveMerge(arrays) {
  const steps = [];
  const combined = [];
  
  arrays.forEach(arr => {
    // Pre-parse timestamps for reliable sorting
    const parsed = arr.map(item => ({ ...item, _time: new Date(item.timestamp).getTime() }));
    combined.push(...parsed);
    
    steps.push({
      type: 'concat',
      desc: `Concatenated ${arr.length} records from ${arr[0]?.satellite} — running total: ${combined.length}`,
      heapState: [],
      extracted: arr[0]?.satellite,
      columns: [],
    });
  });

  steps.push({
    type: 'sort',
    desc: `Sorting ${combined.length} records using JavaScript Array.sort (TimSort) — O(N log N)`,
    heapState: [],
    extracted: null,
    columns: [],
  });

  // Sort by pre-parsed time
  const result = combined.sort((a, b) => a._time - b._time);
  
  // Clean up internal tracking property
  result.forEach(r => delete r._time);

  steps.push({
    type: 'done',
    desc: `✅ Naive merge complete! ${result.length} records sorted chronologically`,
    heapState: [],
    extracted: null,
    columns: [],
  });

  return { result, steps };
}

// ──────────────────────────────────────────────
// Main Merge Dispatcher
// ──────────────────────────────────────────────
function runMerge(arrays, algorithm) {
  const start = performance.now();
  let output;

  switch (algorithm) {
    case 'heap':  output = heapMerge(arrays); break;
    case 'naive': output = naiveMerge(arrays); break;
    default:      output = kWayMerge(arrays);  break;
  }

  const elapsed = +(performance.now() - start).toFixed(3);
  return { ...output, elapsed, algorithm };
}
